### ISSUE TITLE
Toot should not inhibit terminal features such as clickable URLs

### ISSUE THREAD

**Issue by j0ni** [ihabunek/toot#247]
Title says it all. 

Another example would be highlighting text for the clipboard.

I'm using toot 0.30.0 in tmux and alacritty fwiw.

**Comment by monotropos** [comment]
With terminator v1.91-1 (and toot v0.30.1) I have clickable URLs and highlighting text goes to clipboard with the aid of the shift key. Maybe something in the terminal's settings?

**Comment by danschwarz** [comment]
I agree with the sentiment and this is a frustrating issue, but it might be unavoidable given the "windowed" TUI and line wrapping. Terminals don't know that the URL is line wrapped because it appears to end before the rightmost column. Maybe there's a way to hint to the terminal the full URL using ANSI escape codes? Doubt it but I'll check. The only workable alternative, i think, is to display a url across the span of 1 or 2 entire rows so the terminal can interpret it correctly.

**Comment by danschwarz** [comment]
> Maybe there's a way to hint to the terminal the full URL using ANSI escape codes?

Aha, there is!

[https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda](https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda)

**Comment by danschwarz** [comment]
[hacking commences]

**Comment by danschwarz** [comment]
This might require someone more familiar with urwid internals and/or Python than I am.  

Creating a clickable hyperlink that will actually work regardless of line-breaks and so on requires an ANSI escape sequence of the form:

`\033]8;id=1;http://example.com\e\\Link Text\033]8;;\e\\`

You can try this with your own terminal in bash shell with this script, to see if your terminal supports the escape sequence.

`printf '\033]8;id=1;http://example.com\e\\This is a link\033]8;;\e\\\n'`

To make this work in toot tui, we'd need a way to wrap the existing Link Text in the above escape sequence (the ID number can be omitted but eventually we'd want to use IDs to identify links that wrap across multiple lines onscreen).  

I thought it might be possible to subclass urwid.widget.Text, creating a URLText class that basically does the same thing as Text, but adds the escape sequence to the start and the end of the text.  Then I changed the timeline references from Text to URLText wherever url links were being output. I couldn't get this to work; the escape sequences are getting munged and the whole thing comes out as plaintext.  

The escape sequences need to be added at at a lower level in urwid, likely the Canvas.content() method.  This perhaps requires reimplementing TextCanvas for use by the URLText class.  Tried this and haven't gotten very far with it yet.

Any suggestions?

**Comment by danschwarz** [comment]
Update: got a proof-of-concept working!

See below.  I have implemented a new widget, URLText, which is mostly identical to the urwid Text widget, and a URLTextCanvas, based on TextCanvas.  These accept an additional url parameter as input.  On output, this widget emits the proper ANSI escape sequence to embed the full URL in the terminal.  As you can see in the photo, the text on screen is truncated with an ellipsis... but the link is clickable and brings up the full URL properly in the web browser.

<img width="1427" alt="itworks" src="https://user-images.githubusercontent.com/3261094/205466385-b8a1b0a8-46a2-4dd5-b5d1-34aa52eafcb4.png">

**Comment by danschwarz** [comment]
Issues to deal with:

The clickable link extends past the link text when the window is wide
Links embedded in Status messages (as opposed to Media links, for example) are not parsed out and rendered as URLLinks, therefore they don't behave any better than 0.30.1.  It's likely this can be fixed. 

I can make this code available as a patch if anyone wants to continue hacking on it.


<img width="1535" alt="longline" src="https://user-images.githubusercontent.com/3261094/205466534-c2dd861e-b0d5-4564-a7e4-77992e783520.png">

**Comment by danschwarz** [comment]
Patch submitted. It now only underlines the actual link text, so it looks and works correctly in a terminal like windows terminal or terminology that supports the protocol. Handles some links embedded in status messages but not all of them. This can be the subject of further code changes.  By the way, this can help with the copy issue as well, if your terminal program has a copy-url feature. I think windows Terminal does. 

[patch link](https://lists.sr.ht/~ihabunek/toot-discuss/patches/37331)

**Comment by danschwarz** [comment]
Finally making progress on a version that supports links "perfectly" with multi-line wrapping and so on. Coming soon to a PR...

![image](https://user-images.githubusercontent.com/3261094/235325074-a9bacd97-b6e5-484e-a4e6-07c808e1362a.png)
